readme.md

```diff
@@ -258,6 +258,24 @@ Determines if a `got.HTTPError` is thrown for error responses (non-2xx status co
 
 If this is disabled, requests that encounter an error status code will be resolved with the `response` instead of throwing. This may be useful if you are checking for resource availability and are expecting error responses.
 
+###### hooks
+
+Type: `Object<string, Array<Function>>`<br>
+Default: `{ beforeRequest: [] }`
+
+Hooks allow modifications during the request lifecycle. Hook functions may be async and are run serially.
+
+###### hooks.beforeRequest
+
+Type: `Array<Function>`<br>
+Default: `[]`
+
+Called with the normalized request options just before the request is sent. You can modify the object. This is especially useful in conjunction with [`got.extend()`](#instances) and [`got.create()`](advanced-creation.md) when you want to create an API client that uses HMAC-signing.
+
+See the [AWS section](#aws) for an example.
+
+**Note**: Modifying the `body` is not recommended because the `content-length` header has already been computed and assigned.
```

Comment:
There are a few ways this could be strictly prohibited, but I think this is case where it's best to provide guidance but not prohibit advanced usage. Someone may have a valid reason for modifying the body, so this note is here to hint they they probably need to reassign `content-length` if they change body.

Reply:
I agree. Or maybe just move

https://github.com/jstewmon/got/blob/c0aa4d74b9d9d94c858520586b5def68440cd9a9/source/request-as-event-emitter.js#L246

right after `try` before it calculates the body size?

Reply:
That would violate the expectation that got will make no further changes to the options before the request is sent. ;-)

Reply:
> That would violate the expectation that got will make no further changes to the options before the request is sent.

But that's only one header... I think that'd be fine :)

@sindresorhus What do you think?

Reply:
If that is a signed header, it will break the whole thing. It's a big deal.

Reply:
I updated to say that Got will make no further changes to the request before it is sent.

I used the signing feature as an example of why I did not want to move the call before the last attempt to set `content-length`, but we shouldn't specifically mention request signing in the docs for `beforeRequest`, since it's not tightly coupled to request signing.